Use inih instead of libiniparser

columbarius opened the pull request:
Resolves: #96 

Note: inih doesn't escape quotes around a value, so this will break some configs.

#118 has made parsing much simpler.

danshick commented:
Can you give an example of broken configs due to quoting? Does the documentation need to be updated? Can we add the escaping?

columbarius commented:
In the example [config](https://github.com/emersion/xdg-desktop-portal-wlr/blob/master/contrib/config.sample#L4) i used
```
chooser_cmd="slurp -f %o -o"
```
which will result now in an error, because sh can't find `"slurp -f %o -o"`.
Those quotes are not necessary and I can't remember any reason, why i added them in the first place.

So the problem is not inih not adding quotes, but libiniparser ignoring some quotes and with a sample recommending this "wrong"/unneeded quotes.

I just updated the example and the documentation with this pr.

emersion commented:
Yeah, I think it's OK to do this breaking change.

emersion commented on src/core/config.c:
```diff
@@ -28,63 +30,57 @@ void finish_config(struct xdpw_config *config) {
 		config->screencast_conf.chooser_type = get_chooser_type(chooser_type);
 		free(chooser_type);
+	} else {
+		logprint(TRACE, "config: skipping invalid key in config file");
```
Making this an error would make it easier for users to understand why one of their config options doesn't work (e.g. typo). This is kind of unrelated to this PR though.

columbarius replied:
This only works nice together with [`logprint(TRACE, "config: parsing setction %s, key %s, value %s", section, name, value);`](https://github.com/emersion/xdg-desktop-portal-wlr/pull/126/files/1b7b22ec2665cdd11c02ae96ec2df79cf6cef8f2#diff-4189891b601c97c24b3f16f7fea437d38af30d2d1c34fc8b692e7b96b153a37cR52)

emersion commented on src/core/config.c:
```diff
@@ -28,63 +30,57 @@ void finish_config(struct xdpw_config *config) {
+	struct xdpw_config *config = (struct xdpw_config*)data;
+	logprint(TRACE, "config: parsing setction %s, key %s, value %s", section, name, value);
+
+	#define MATCH(s, n) strcmp(section, s) == 0 && strcmp(name, n) == 0
```
This can be a function, bringing type safety in.

emersion replied:
If the `section` repetition is too verbose, we could have one handler per section:

```c
static int handle_ini_screencast(struct xdpw_config *config, const char *name, const char *value);
```

emersion replied:
About macros: https://drewdevault.com/2017/03/15/How-I-learned-to-stop-worrying-and-love-C.html

emersion commented on src/core/config.c:
```diff
@@ -28,63 +30,57 @@ void finish_config(struct xdpw_config *config) {
-	getstring_from_conffile(d, "screencast:exec_after", &config->screencast_conf.exec_after, NULL);
-	getstring_from_conffile(d, "screencast:chooser_cmd", &config->screencast_conf.chooser_cmd, NULL);
-	if (!config->screencast_conf.chooser_type) {
+static int ini_parser_handler(void *data, const char* section, const char* name, const char* value) {
```
Style: `*` on the right

emersion reviewed (approved):
LGTM, thanks!

columbarius commented:
Thanks for review and merging
